pages/blog/varchar-191.mdx

```diff
@@ -0,0 +1,78 @@
+---
+title: "Why do database columns have a character length of 191?"
+date: "2021-05-14"
+author: "Evan Tahler"
+pullQuote: "`varchar(191)` is such an odd number - how did we get here?"
+image: "varchar-191/210515-varchar191.png"
+tags: [engineering, notes]
+---
+
+<Image
+  centered
+  alt="A Database symbol over a library"
+  src="varchar-191/210515-varchar191.png"
+  width={1015}
+  height={533}
+/>
+
+Sometimes, when you are looking at a database’s schema, you see that there are text fields defined like this:
+
+```sql
+email_address varchar(191) NOT NULL
+```
+
+This means that the column supports strings with a maximum length of 191 characters, and can’t be `null`. 191 is such an odd number - where did it come from? In this post, we’ll look at the historical reasons for the 191 character limit as a default in most relational databases.
+
+## Why `varchar` and not `text`?
+
+The first question you might ask is why limit the length of the strings you can store in a database at all? All modern popular relational database support (almost) unlimited sized strings with a `text` or `blob`-type column, so why not use that? The reason is **indexes**.
+
+If you are going to search by a column, say `email_address`, you probably want to add an index to it to speed things up when you do the following:
+
+```sql
+select id from users where email = 'foo@example.com';
+```
+
+As your table gets bigger, searches get slower because your database has to check every row to find a match. However, if you add a **search index**, you are telling your database to essentially "pre-compute" popular search patterns with a tree so the next search is much faster. In essence, indexes spend computation time (and a little bit of disk space) making writes to the database slower, to speed up reads later. For most applications this is a great tradeoff, since they are "read heavy" and "write lite".
+
+So, why use `varchar`? Indexes can be made to preform better when assumptions can be made about the type of data they store. Knowing how long the strings in the index are is one of the best ways speed things up. For some databases, you aren’t allowed to add a search index to columns of type `text` becaus this optimization can't be done, while in others, the index just won’t preform as well. In fact, historically, databases were constructed with limits on how big an index could be to optimize search and how they stored data on disk.
```

Comment:
```suggestion
So, why use `varchar`? Indexes can be made to perform better when assumptions can be made about the type of data they store. Knowing how long the strings in the index are is one of the best ways speed things up. For some databases, you aren’t allowed to add a search index to columns of type `text` because this optimization can't be done, while in others, the index just won’t perform as well. In fact, historically, databases were constructed with limits on how big an index could be to optimize search and how they stored data on disk.
```